File: donghwi/week23/BAEK_17413_단어 뒤집기2.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader((System.in)));

        StringBuilder word = new StringBuilder();
        StringBuilder result = new StringBuilder();

        // 태그 안에 있는지 확인하는 플래그
        boolean inTag = false; 

        for (char c : br.readLine().toCharArray()) {

            if (c == '<') {
                inTag = true;
                // 태그가 시작되면 그동안 쌓인 단어를 뒤집어서 결과에 추가
                result.append(word.reverse());
                word.setLength(0);
                result.append(c);
            } else if (c == '>') {
                inTag = false;
                result.append(c);
            } else if (inTag) {
                // 태그 안에서는 그대로 추가
                result.append(c);
            } else {
                // 태그 밖일 때
                if (c == ' ') {
                    // 공백을 만나면 단어를 뒤집어 추가하고 공백도 추가
                    result.append(word.reverse());
                    result.append(c);
                    word.setLength(0);
                } else {
                    // 단어의 일부라면 뒤집기 위해 임시 보관
                    word.append(c);
                }
            }
        }
        // 마지막에 남은 단어 처리
        result.append(word.reverse());

        System.out.println(result);
    }
}

File: youngcheol/20260128/BAEK_17413_단어뒤집기2.py
# 알파벳, 숫자 < >로 구성
# < > 안에 들어간 문자열은 뒤집지 말고 
# 나머지는 다 뒤집는다.


S = input()

stack = []
# 결과 저장 
result = []
in_tag = False

for i in S:
    if i == '<':
        # <을 만나게 되면 stack에 있는 말 다 뒤집기
        while stack:
            print(stack.pop(), end='')

        # < > 안에 있는 말은 어떻게 해야할지 몰라서 제미나이 도움 받음
        in_tag = True
        print(i, end ='')

    # >이면 그대로 출력
    elif i == '>':
        in_tag = False
        print(i, end='')
    
    # 태그 안에 있는 문자열 그대로 출력
    elif in_tag:
        print(i, end ='')
        
    else:
        if i == ' ':
            while stack:

                print(stack.pop(), end='')
            print(' ', end='')
        else:
            stack.append(i)

while stack:
    print(stack.pop(), end = '')


